### PR TITLE
[WIP] - creating sources (ElastichSearch, mysql, redis, etc) connectors

### PR DESCRIPTION
**Pull request by eavilesmejia**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "license": "MIT",
     "require": {
         "php": ">=7.0.0",
-        "nesbot/carbon": "1.21.0"
+        "nesbot/carbon": "1.21.0",
+        "elasticsearch/elasticsearch": "dev-master"
```
Comment (iampersistent) [inline composer.json]: ES shouldn't be required. It should go int require-dev and suggested

File: tests/unit/ConnectorTransformCest.php [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+                    'name'       => 'My name is Elizabeth',
+                    'age'        => 42,
+                    'occupation' => [
+                        'name'      => 'Waitress',
+                        'startDate' => '2017-01-01 10:10:10',
+                    ],
+                ],
+            ],
+        ];
+        $output = (new TransformService())->untransform($payload, new PersonTransformer());
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: These need to be updated to ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+                    ],
+                ],
+            ],
+        ];
+        $I->assertEquals($expected, $output);
+    }
+
+    public function testTransformationWithoutKey(UnitTester $I)
+    {
+        $output = (new TransformService())->transform(new Person('Bill', 'Piano Man'), new PersonTransformer());
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: Update to ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+    public function testTransformationCollectionAll(UnitTester $I)
+    {
+        $friends = [
+            new Person('Paul', 'Real estate novelist'),
+            new Person('John', 'Bartender'),
+            new Person('Davy', 'Sailor'),
+            new Person('Elizabeth', 'Waitress'),
+        ];
+        $output = (new TransformService())->transform($friends, new PersonTransformer(), 'people', '*',
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: Update to ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+    public function testTransformationCollectionWithoutKey(UnitTester $I)
+    {
+        $friends = [
+            new Person('Paul', 'Real estate novelist'),
+            new Person('John', 'Bartender'),
+            new Person('Davy', 'Sailor'),
+            new Person('Elizabeth', 'Waitress'),
+        ];
+        $output = (new TransformService())->transform($friends, new PersonTransformer(), '', '*',
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        $friends = [
+            new Person('Paul', 'Real estate novelist'),
+            new Person('John', 'Bartender'),
+            new Person('Davy', 'Sailor'),
+            new Person('Elizabeth', 'Waitress'),
+        ];
+
+        $fieldMap = new FieldMap(['name' => '*', 'my_job' => ['name']]);
+
+        $output = (new TransformService())->transform($friends, new PersonTransformer(), 'people', $fieldMap,
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: Update to ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        $friends = [
+            new Person('Paul', 'Real estate novelist'),
+            new Person('John', 'Bartender'),
+            new Person('Davy', 'Sailor'),
+            new Person('Elizabeth', 'Waitress'),
+        ];
+
+        $fieldMap = new FieldMap(['name', 'my_job']);
+
+        $output = (new TransformService())->transform($friends, new PersonTransformer(), 'people', $fieldMap,
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: Update to ConnectorTransformService

File: tests/unit/ConnectorTransformCest.php [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+                    'my_job' => [
+                        'name'       => 'Waitress',
+                        'start_date' => '2017-01-01 10:10:10',
+                    ],
+                ],
+            ],
+        ];
+        $I->assertEquals($expected, $output);
+    }
+
```
Comment (iampersistent) [inline tests/unit/ConnectorTransformCest.php]: There also needs to be a test for untransforming from array to object

File: composer.json (hunk 1 above)
Comment (iampersistent) [inline composer.json]: ES shouldn't be required. It should go int require-dev and suggested

File: src/ConnectorTransformService.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    public function transform(Connector $connector, AbstractTransformer $transformer, $key = '', $fieldMap = '*', $resourceClass = Object::class)
+    {
+        $resource = new $resourceClass($connector->getData(), '', $fieldMap);
+        $transformed = $transformer->transform($resource);
+
+        if ($resource instanceof Collection) {
+            $connector->sendDataToSource($transformed);
+        } else {
+            $connector->sendDataToSource($transformed);
```
Comment (alexcoffee) [inline src/ConnectorTransformService.php]: I don't see the difference, is that just there for future support?

Comment (eavilesmejia) [inline src/ConnectorTransformService.php]: Yep, we haven't finished the connector approach yet

**Pull request closed without merging**